openlibrary/core/lending.py

```diff
@@ -157,13 +157,18 @@ def compose_ia_url(limit=None, page=1, subject=None, query=None, work_id=None,
                         authors.append(author_name)
                         authors.append(','.join(author_name.split(' ', 1)[::-1]))
                     if authors:
-                        _q = ' OR '.join(['creator:"%s"' % author for author in authors])
+                        _q = ' OR '.join('creator:"%s"' % author for author in authors)
                 elif _type == "subjects":
                     subjects = works_authors_and_subjects.get('subjects', [])
                     if subjects:
-                        _q = ' OR '.join(['subject:"%s"' % subject for subject in subjects])
+                        _q = ' OR '.join('subject:"%s"' % subject for subject in subjects)
             if not _q:
-                return []
+                logger.error('compose_ia_url(limit={}, page={}, subject={}, query={}, '
+                    'work_id={}, _type={}, sorts={}, advanced={}) failed!'.format(
+                        limit, page, subject, query, work_id, _type, sorts, advanced
+                    )
+                )
+                return ''  # TODO: Should we just raise an excpetion instead?
```

Comment:
Just commentary, not requesting a change:

It looks to me the primary consumer of this function (`compose_ia_url`) is this carousel: https://github.com/internetarchive/openlibrary/blob/698d08fd4eda5ce76606837f89cbfffe2445b106/openlibrary/templates/home/custom_ia_carousel.html

I don't have a strong opinion on exception vs. return empty string, but either way, it seems to me that the carousel ought to have its behavior change in this situation. Is the smartest thing to remove the book from the carousel if the we can't find an IA URL? If so, perhaps it makes sense (in a follow-up PR) to adjust https://github.com/internetarchive/openlibrary/blob/698d08fd4eda5ce76606837f89cbfffe2445b106/openlibrary/templates/home/custom_ia_carousel.html so that if the `url` it gets is the empty string, then it returns the empty string rather than returns the template?

Also, typo in `excpetion`; should be `exception`.

Out of curiosity, how does someone read these logs? Presumably by SSH-ing into prod and grepping a log file? That's fine, just curious.

Perhaps it makes sense to keep an issue open to track finding the data from the logs and fixing this function so we can eventually find and fix all those cases.

Related question: where does this bad data even come from? Does it make sense to treat this case as if L149 `if work_id` is false? Then rather than returning, we would log the error, and continue down.

Reply:
:+1: Created one here, and put a log dump in it #3429.

> Related question: where does this bad data even come from? Does it make sense to treat this case as if L149 if work_id is false? Then rather than returning, we would log the error, and continue down.

I think we'll need some of the log data to determine why it's coming down this path at all :/ It might be a bug in upstream code that's causing maybe `cached_work_authors_and_subjects` to fail? Or it could be being passed in a wrong parameter... 🤷